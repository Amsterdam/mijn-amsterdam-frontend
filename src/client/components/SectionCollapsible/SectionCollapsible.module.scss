@import '../../styles/_global.scss';

.SectionCollapsible {
  margin-bottom: 1.8rem;
  border-bottom: 0.1rem solid $color-neutral-grey2;
}

@media print {
  .SectionCollapsible {
    break-before: always;
  }
}

.CaretIcon {
  height: 0.8em;
  width: 0.8em;
  position: absolute;
  top: 50%;
  transform: translateY(calc(-50% - 1px)) rotate(0);
  left: 0;
  transition: transform 200ms ease-out;
}

.SectionCollapsible .Title {
  position: relative;
  padding-left: 0.96em;
  margin-left: 16rem;
  margin-bottom: 1.8rem;
  transform: translateX(-2.2rem);
  white-space: nowrap;

  @include mq-tablet() {
    margin-left: 9.5rem;
    white-space: initial;
  }

  @include mq-phone() {
    margin-left: 0;
    transform: translateX(0);
  }

  span {
    padding-left: 0.2em;
  }
}

.TitleToggle {
  margin: 0;
  border: 0;
  background: 0;
  font-weight: bold;
  padding: 0;
  text-align: left;
  color: inherit;

  &[aria-expanded='true'] .CaretIcon {
    transform: translateY(-50%) rotate(90deg);
  }
}

.Panel {
  overflow: hidden;
  height: 0;
  position: relative;
}

@media print {
  .Panel {
    overflow: visible;
    height: auto !important;
  }
}

.NoItemsMessage {
  @include mq-tablet() {
    text-indent: 12rem;
  }

  @include mq-phone() {
    text-indent: 0;
  }

  text-indent: 16rem;
  margin-top: 1rem;
  padding-bottom: 3rem;
}

.ToggleMorePanel {
  background: $color-white;
  position: absolute;
  bottom: 0;
  left: 0;
  right: 0;
}

.PanelInner {
  display: block;
}
