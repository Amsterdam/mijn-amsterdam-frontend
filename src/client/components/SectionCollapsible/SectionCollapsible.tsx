import classnames from 'classnames';
import React, { ReactNode, useEffect, useMemo, useRef, useState } from 'react';
import { animated, useSpring } from 'react-spring';
import { useDebouncedCallback } from 'use-debounce';
import { withKeyPress } from '../../../universal/helpers';
import { ComponentChildren } from '../../../universal/types';
import { IconChevronRight } from '../../assets/icons';
import { trackEventWithProfileType, useSessionStorage } from '../../hooks';
import { useComponentSize } from '../../hooks/useComponentSize';
import { useProfileTypeValue } from '../../hooks/useProfileType';
import Heading from '../Heading/Heading';
import LoadingContent from '../LoadingContent/LoadingContent';
import styles from './SectionCollapsible.module.scss';

export interface SectionCollapsibleProps {
  id: string;
  title?: string | ReactNode;
  noItemsMessage?: string;
  startCollapsed?: boolean;
  className?: string;
  isLoading?: boolean;
  track?: { category: string; name: string };
  hasItems?: boolean;
  children: ComponentChildren;
}

export interface SectionCollapsibleHeadingProps {
  children: ComponentChildren;
  toggleCollapsed: (event: any) => void;
  isAriaExpanded: boolean;
}

export function SectionCollapsibleHeading({
  children,
  toggleCollapsed,
  isAriaExpanded,
}: SectionCollapsibleHeadingProps) {
  return (
    <Heading
      size="mediumLarge"
      className={classnames(styles.Title, styles.TitleWithItems)}
    >
      <button
        aria-expanded={isAriaExpanded}
        className={styles.TitleToggle}
        onKeyPress={(event) => toggleCollapsed(event)}
        onClick={(event) => toggleCollapsed(event)}
      >
        <IconChevronRight aria-hidden="true" className={styles.CaretIcon} />{' '}
        {children}
      </button>
    </Heading>
  );
}

interface SectionCollapsibleBodyProps {
  children: ComponentChildren;
  className?: string;
}

export function SectionCollapsibleBody({
  children,
  className,
}: SectionCollapsibleBodyProps) {
  const classes = classnames(styles.SectionCollapsible, className);
  return <section className={classes}>{children}</section>;
}

export default function SectionCollapsible({
  id,
  title = '',
  noItemsMessage = '',
  startCollapsed = true,
  className,
  isLoading = false,
  hasItems = true,
  track,
  children,
}: SectionCollapsibleProps) {
  const [isCollapsed, setCollapsed] = useSessionStorage(id, startCollapsed);
  const contentRef = useRef<HTMLDivElement>(null);
  const [isReadyForAnimation, setReadyForAnimaton] = useState(false);
  const hasTitle = !!title;
  const hasNoItemsMessage = !!noItemsMessage;
  const profileType = useProfileTypeValue();

  const setReadyForAnimatonDebounced = useDebouncedCallback(() => {
    if (!isLoading && isReadyForAnimation === false) {
      setReadyForAnimaton(true);
    }
  }, 50);

  setReadyForAnimatonDebounced.callback();

  const [{ height: contentHeight }, setDimensions] = useState({
    width: 0,
    height: 0,
  });

  const size = useComponentSize(contentRef.current);

  const contentDimensions = useMemo(() => {
    return {
      width: size.width,
      height: size.height,
    };
  }, [size.height, size.width]);

  useEffect(() => {
    if (!isLoading && contentRef && contentRef.current) {
      setDimensions(contentDimensions);
    }
  }, [isLoading, contentDimensions]);

  const toggleCollapsed = withKeyPress<HTMLSpanElement>(() => {
    if (isCollapsed && track) {
      trackEventWithProfileType(
        {
          ...track,
          action: 'Open klikken',
        },
        profileType
      );
    }
    setCollapsed(!isCollapsed);
  });

  const heightAnim = {
    immediate: !isReadyForAnimation,
    reverse: isCollapsed,
    from: {
      height: 0,
    },
    height: contentHeight,
  };

  const heightAnimSpring = useSpring(heightAnim);

  return (
    <SectionCollapsibleBody className={className}>
      {hasTitle && (
        <SectionCollapsibleHeading
          isAriaExpanded={!isCollapsed}
          toggleCollapsed={toggleCollapsed}
        >
          {title}
        </SectionCollapsibleHeading>
      )}
      {isLoading && (
        <LoadingContent
          barConfig={[
            ['auto', '2rem', '1rem'],
            ['auto', '2rem', '0'],
          ]}
        />
      )}

      <animated.div
        aria-hidden={isCollapsed}
        className={styles.Panel}
        style={heightAnimSpring}
      >
        <div className={styles.PanelInner} ref={contentRef}>
          {hasNoItemsMessage && !isLoading && !hasItems && (
            <p className={styles.NoItemsMessage}>{noItemsMessage}</p>
          )}
          {hasItems && children}
        </div>
      </animated.div>
    </SectionCollapsibleBody>
  );
}
