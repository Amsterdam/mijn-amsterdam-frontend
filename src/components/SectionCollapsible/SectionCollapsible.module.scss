@import 'styles/_global.scss';

.SectionCollapsible {
  overflow: auto; // fixes margin of wrapped SectionCollapsible.Title being respected

  &.isCollapsed {
    margin-bottom: 0;
    .CaretIcon {
      transform: translateY(-50%) rotate(0);
    }
  }
}

.CaretIcon {
  height: 0.8em;
  width: 0.8em;
  position: absolute;
  top: 1.5rem;
  transform: translateY(-50%) rotate(90deg);
  left: 0;
  transition: transform 200ms ease-out;
}

.Title {
  position: relative;
  padding-left: 1em;
  margin: 0 auto;
  margin-bottom: 3rem;

  @include mq-tablet() {
    margin-left: 9.5rem;
  }
  @include mq-phone() {
    margin-left: 0;
  }
  max-width: 62rem;

  &WithItems span {
    cursor: pointer;
  }

  span {
    padding-left: 0.2em;
  }
}

.Panel {
  overflow: hidden;
  border-bottom: 0.1rem solid $color-neutral-grey2;
  height: 0px;

  @include mq-phone() {
    border-bottom: 0;
  }
}

.NoItemsMessage {
  @include mq-tablet() {
    text-indent: 12rem;
  }
  @include mq-phone() {
    text-indent: 0;
  }
  text-indent: 18rem;
  margin-top: -3rem; // closer to the heading
  margin-bottom: 3rem;
}


:global {
  .sectionCollapsible-enter-active,
  .sectionCollapsible-exit-active,
  .sectionCollapsible-enter-done,
  .sectionCollapsible-exit-done,
  .sectionCollapsible-exit,
  .sectionCollapsible-enter {
    transition: all 300ms;
  }
}
