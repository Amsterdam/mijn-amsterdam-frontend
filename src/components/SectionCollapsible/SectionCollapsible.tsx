import { ComponentChildren } from 'App.types';
import { ReactComponent as CaretIcon } from 'assets/icons/Chevron-Right.svg';
import classnames from 'classnames';
import LoadingContent from 'components/LoadingContent/LoadingContent';
import { withKeyPress } from 'helpers/App';
import { useSessionStorage } from 'hooks/storage.hook';
import React, { useEffect, useState } from 'react';

import Heading from '../Heading/Heading';
import styles from './SectionCollapsible.module.scss';
import { trackEvent } from 'hooks/analytics.hook';
import { useRef } from 'react';
import { useDomElementDimensions } from 'hooks/useDomElementDimensions.hook';
import { useSpring, animated } from 'react-spring';
import { useDebouncedCallback } from 'use-debounce';

export interface SectionCollapsibleProps {
  title?: string;
  noItemsMessage?: string;
  isCollapsed: boolean;
  onToggleCollapsed: () => void;
  className?: string;
  isLoading?: boolean;
  track?: { category: string; name: string };
  hasItems?: boolean;
  children: ComponentChildren;
}

export default function SectionCollapsible({
  title = '',
  noItemsMessage = '',
  isCollapsed,
  onToggleCollapsed,
  className,
  isLoading = false,
  hasItems = true,
  track,
  children,
}: SectionCollapsibleProps) {
  const contentRef = useRef(null);
  const [isReadyForAnimation, setReadyForAnimaton] = useState(false);
  const { height: contentHeight } = useDomElementDimensions(contentRef);
  const hasTitle = !!title;
  const hasNoItemsMessage = !!noItemsMessage;

  const [setReadyForAnimatonDebounced] = useDebouncedCallback(() => {
    if (!isLoading) {
      setReadyForAnimaton(true);
    }
  }, 200);

  setReadyForAnimatonDebounced();

  const classes = classnames(
    styles.SectionCollapsible,
    (!hasItems || isCollapsed) && styles.isCollapsed,
    className
  );

  const toggleCollapsed = withKeyPress<HTMLSpanElement>(() => {
    if (isCollapsed && track) {
      trackEvent({
        ...track,
        action: 'Open klikken',
      });
    }
    onToggleCollapsed && onToggleCollapsed();
  });

  const heightAnim = {
    immediate: !isReadyForAnimation,
    reverse: isCollapsed,
    from: {
      height: 0,
    },
    height: contentHeight + 2,
  };

  const heightAnimSpring = useSpring(heightAnim);

  return (
    <section className={classes}>
      {hasTitle && (
        <Heading
          size="mediumLarge"
          className={classnames(
            styles.Title,
            hasItems && styles.TitleWithItems
          )}
        >
          <CaretIcon aria-hidden="true" className={styles.CaretIcon} />{' '}
          <span
            {...(hasItems ? { role: 'button', tabIndex: 0 } : {})}
            onKeyPress={event => hasItems && toggleCollapsed(event)}
            onClick={event => hasItems && toggleCollapsed(event)}
          >
            {title}
          </span>
        </Heading>
      )}
      {isLoading && (
        <LoadingContent
          barConfig={[['auto', '2rem', '1rem'], ['auto', '2rem', '0']]}
        />
      )}
      {hasNoItemsMessage && !isLoading && !hasItems && (
        <p className={styles.NoItemsMessage}>{noItemsMessage}</p>
      )}

      {!isLoading && hasItems && (
        <animated.div
          aria-hidden={isCollapsed}
          className={styles.Panel}
          style={heightAnimSpring}
        >
          <div ref={contentRef}>{children}</div>
        </animated.div>
      )}
    </section>
  );
}
