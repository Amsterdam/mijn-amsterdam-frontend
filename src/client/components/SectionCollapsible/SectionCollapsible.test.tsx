import { mount, ReactWrapper } from 'enzyme';
import React from 'react';
import * as analytics from '../../hooks/analytics.hook';
import SectionCollapsible from './SectionCollapsible';

describe('SectionCollapsible', () => {
  let component: ReactWrapper<typeof SectionCollapsible>;
  const trackingSpy = jest.spyOn(analytics, 'trackEvent');

  afterEach(() => {
    component.unmount();
    sessionStorage.clear();
  });

  it('should start uncollapsed', () => {
    component = mount(
      <SectionCollapsible
        id="test-SectionCollapsible"
        isLoading={false}
        hasItems={true}
        startCollapsed={false}
        title="test!"
      >
        <div style={{ height: 500 }}>Boohoo!</div>
      </SectionCollapsible>
    );
    expect(
      (component.find('SectionCollapsibleHeading').props() as any)
        .isAriaExpanded
    ).toBe(true);
  });

  it('should start collapsed', () => {
    component = mount(
      <SectionCollapsible
        id="test-SectionCollapsible"
        isLoading={false}
        hasItems={true}
        title="test"
      >
        <div style={{ height: 500 }}>Boohoo!</div>
      </SectionCollapsible>
    );

    expect(
      (component.find('SectionCollapsibleHeading').props() as any)
        .isAriaExpanded
    ).toBe(false);
  });

  it('should collapse/expand when clicking the title', () => {
    component = mount(
      <SectionCollapsible
        id="test-SectionCollapsible"
        isLoading={false}
        hasItems={true}
        title="Click me!"
      >
        <div style={{ height: 500 }}>Boohoo!</div>
      </SectionCollapsible>
    );

    expect(
      (component.find('SectionCollapsibleHeading').props() as any)
        .isAriaExpanded
    ).toBe(false);
    component.find(`.Title button`).simulate('click');
    expect(
      (component.find('SectionCollapsibleHeading').props() as any)
        .isAriaExpanded
    ).toBe(true);
  });

  it('should call trackEvent if tracking info is provided and section is expanded', () => {
    component = mount(
      <SectionCollapsible
        id="test-SectionCollapsible"
        isLoading={false}
        hasItems={true}
        track={{ category: 'the category', name: 'the content thing' }}
        title="Click me!"
      >
        <div style={{ height: 500 }}>Boohoo!</div>
      </SectionCollapsible>
    );

    component.find(`.Title button`).simulate('click');
    expect(trackingSpy).toHaveBeenCalledWith({
      category: 'the category',
      name: 'the content thing',
      action: 'Open klikken',
    });
  });

  it('should show title and "no items message"', () => {
    component = mount(
      <SectionCollapsible
        id="test-SectionCollapsible"
        title="My Items"
        isLoading={false}
        hasItems={false}
        noItemsMessage="No items"
      >
        <div style={{ height: 500 }}>Boohoo!</div>
      </SectionCollapsible>
    );

    expect(component.html()).toMatchSnapshot();
  });
});
