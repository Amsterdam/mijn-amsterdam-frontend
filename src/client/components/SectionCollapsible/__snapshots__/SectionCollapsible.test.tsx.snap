// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionCollapsible should show title and "no items message" 1`] = `"<section class=\\"SectionCollapsible\\"><h3 class=\\"Heading Heading__mediumLarge Title\\"><svg aria-hidden=\\"true\\" class=\\"CaretIcon\\">Chevron-Right.svg</svg> My Items</h3><p class=\\"NoItemsMessage\\">No items</p></section>"`;
